content/en/blog/_posts/advanced-server-side-apply.md

```diff
@@ -0,0 +1,80 @@
+
+The only downside is that you may end up sending unneeded APPLY requests to apiserver, if actually the object already matches your controller’s desires. This doesn’t matter if it happens once in a while, but for extremely high-throughput controllers, it might cause a performance problem for the cluster–specifically, apiserver. No-op writes are not written to storage (etcd) or broadcast to any watchers, so it’s not really that big of a deal. If you’re worried about this anyway, today you could use the method explained in the previous section, or you could still do it this way for now, and wait for an additional client-side mechanism to suppress zero-change suppress.
+
+To get around this downside, why not GET the object and only send your APPLY if the object needs it? Surprisingly, it doesn’t help much– a no-op APPLY is not very much more work for apiserver than an extra GET; and an APPLY that changes things is cheaper than that same APPLY with a preceding GET. Worse, since it is a distributed system, something could change between your GET and APPLY, invalidating your computation. Instead, we can use this optimization on an object retrieved from a cache–then it legitimately will reduce load on the system (at the cost of a delay when a change is needed and the cache is a bit behind).
+
+#### CI/CD systems
+Continuous integration and/or deployment systems are a special kind of controller which is doing something like reading source .yaml files from a git repo and automatically pushing them into the cluster, perhaps after running them through some testing, or waiting for test results, etc. Generally users are the entities pushing changes into the git repo, although that’s not necessarily always the case.
+
+Some systems like this continuously reconcile with the cluster, others may only operate when a change is pushed to the git repo. The following considerations are important for both, but more so for the continuously reconciling kind.
+
+CI/CD systems are literally controllers, but for the purpose of APPLY, they are more like users, and unlike other controllers, they need to pay attention to conflicts. Reasoning:
+* Abstractly, CI/CD systems can change anything, which means they could conflict with any controller out there. The recommendation that controllers force conflicts is assuming that controllers change a limited number of things and you can be reasonably sure that they won’t fight with other controllers about those things; that’s clearly not the case for CI/CD controllers.
+* Concrete example: imagine the CI/CD system wants spec.replicas to be 3, because that is checked into source code; but HPA is also enabled for this Deployment, and it thinks there should be 10 replicas. Who should win? We just said that most controllers–including the HPA–should ignore conflicts. The HPA has no idea if it has been enabled incorrectly, and the HPA has no convenient way of informing users of errors.
```

Comment:
```suggestion
* Concrete example: imagine the CI/CD system wants `.spec.replicas` for some Deployment to be 3, because that is the value that is checked into source code; however there is also a HorizontalPodAutoscaler (HPA) that targets the same deployment. The HPA computes a target scale and decides that there should be 10 replicas. Which should win? I just said that most controllers–including the HPA–should ignore conflicts. The HPA has no idea if it has been enabled incorrectly, and the HPA has no convenient way of informing users of errors.
```